### ISSUE TITLE
Rule proposal: `prefer-array-map`

### ISSUE THREAD

**Issue by Pyrolistical** [sindresorhus/eslint-plugin-unicorn#1712]
### Description

I often find that the [`no-for-loop`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-for-loop.md) results in something that could be simplified down to usage of [`Array.prototype.map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)

### Fail

```js
const result = []
for (const element of array) {
  result.push(transform(element))
}
```
```js
const result = []
for (const [index, element] of array.entries()) {
  result.push(anotherArray[index](element))
}
```

### Pass

```js
const result = array.map((element) => transformer(element))
```
```js
const result = array.map((element, index) => anotherArray[index](element))
```

**Comment by Pyrolistical** [comment]
Hmm, this rule could be problematic for stuff beyond arrays. for..of isn't just for arrays, its generally for iterables https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols

So this transformation is only valid if we know `x` in `for (... of x) {` is an Array

**Comment by sindresorhus** [comment]
I would wait on https://github.com/tc39/proposal-iterator-helpers

**Comment by Pyrolistical** [comment]
@fisker interesting. maybe that's a different rule, `prefer-array-from` and uses https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

the neat thing about that is it has a built in map function, so it already works for both arrays and iterators

**Comment by silverwind** [comment]
I think I'm against this. `for..of` is generally superiour because it allows flow control statements. `Array#map` has its uses for simple one-liners, but I would not recommend it in the general sense as a replacement of `for..of`.

**Comment by fisker** [comment]
Found one in our codebase

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/ba9d53cd522e641d9b8da7ce495ed0110b8ce33a/test/package.mjs#L100-L111

**Comment by B2o5T** [comment]
Another case if there is more than 1 argument in `Array#push` method, should be reported by `prefer-array-flat-map`.

**Comment by fregante** [comment]
One exception: when the loop content accesses the array for other reasons (e.g. to look back). Random example:

```js
const result = [];
for (const item of array) {
	if (result.length % 2) {
		result.push('skipped')
	} else {
		result.push(item * 2)
	}
}
```

**Comment by Pyrolistical** [comment]
@fregante yep, the rule wouldn't apply in this case as it access the result before `.map` completes. However, that example could be rewritten as an `.reduce`

**Comment by fregante** [comment]
> I would only allow this rule to match only "simple" `for..of` cases as in these examples.

Maybe a single condition could also be wrapped. From:

```js
const result = []
for (const element of array) {
	if (element > 4)
		result.push(transform(element))
}
```

to:

```js
const result = array
	.filter(element => element > 4)
	.map((element) => transform(element))
```
